File: src/lib/api/export.ts
/**
 * Export: CSV, JSON, PDF, SBOM, STIX, PCAP, firewall rules.
 */

import type { FilteredPcapResult, AllowlistEntry } from '$lib/types/analysis';
import type { EnforcementFormat, SegmentationReport } from '$lib/types/segmentation';
import { z } from 'zod';
import { allowlistEntrySchema, filteredPcapResultSchema } from '$lib/schemas';
import { httpValidated } from './core';
import type { ReportConfig } from '$lib/types';

export async function exportAssetsCsv(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/assets/csv', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function exportConnectionsCsv(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/connections/csv', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function exportTopologyJson(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/topology/json', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function exportAssetsJson(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/assets/json', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function generatePdfReport(config: ReportConfig, outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/report/pdf', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ config, outputPath })
	});
}

export async function exportSbom(format: 'csv' | 'json', outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/sbom', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ format, outputPath })
	});
}

export async function exportStixBundle(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/stix', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function saveTopologyImage(imageData: string, outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/topology/image', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ imageData, outputPath })
	});
}

export async function generateCommunicationAllowlist(): Promise<AllowlistEntry[]> {
	return httpValidated(z.array(allowlistEntrySchema), '/api/v1/exports/allowlist');
}

export async function exportAllowlistCsv(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/allowlist/csv', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function exportFirewallRules(outputPath: string): Promise<string> {
	return httpValidated(z.string(), '/api/v1/exports/firewall-rules', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ outputPath })
	});
}

export async function exportFilteredPcap(
	filterIps: string[],
	filterPorts: number[],
	outputPath: string
): Promise<FilteredPcapResult> {
	return httpValidated(filteredPcapResultSchema, '/api/v1/exports/pcap/filtered', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({
			filterIps,
			filterPorts,
			outputPath
		})
	});
}

export async function runSegmentation(): Promise<SegmentationReport> {
	return httpValidated(z.unknown(), '/api/v1/segmentation/run', { method: 'POST' }) as Promise<SegmentationReport>;
}

export async function exportEnforcementConfig(format: EnforcementFormat): Promise<string> {
	return httpValidated(z.string(), '/api/v1/segmentation/enforcement-config', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ format })
	});
}

File: backend/src/commands/export.rs
//! Adapter layer for export/reporting use-cases.

use gm_analysis::AllowlistEntry;

use crate::application::use_cases::export as use_case;

use super::{
    error::AppError,
    support::{mutex_state, read_state, write_text_file},
    AppState,
};

pub use use_case::file_exports::FilteredPcapResult;
pub use use_case::ReportConfigInput;

pub async fn export_assets_csv(output_path: String, state: &AppState) -> Result<String, AppError> {
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let csv = use_case::export_assets_csv(&inventory.assets).map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &csv).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn export_connections_csv(
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let capture = read_state(&state.capture, "capture").map_err(AppError::state_lock)?;
    let csv =
        use_case::export_connections_csv(&capture.connections).map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &csv).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn export_topology_json(
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let capture = read_state(&state.capture, "capture").map_err(AppError::state_lock)?;
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let session = mutex_state(&state.session, "session").map_err(AppError::state_lock)?;
    let json = use_case::export_topology_json(
        &inventory.assets,
        &capture.connections,
        session.current_session_name.as_deref(),
    )
    .map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &json).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn export_assets_json(output_path: String, state: &AppState) -> Result<String, AppError> {
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let json = use_case::export_assets_json(&inventory.assets).map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &json).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn generate_pdf_report(
    config: ReportConfigInput,
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let capture = read_state(&state.capture, "capture").map_err(AppError::state_lock)?;
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let session = mutex_state(&state.session, "session").map_err(AppError::state_lock)?;
    use_case::generate_pdf_report(
        config,
        output_path,
        &inventory.assets,
        &capture.connections,
        session.current_session_name.as_deref(),
    )
    .map_err(AppError::invalid_input)
}

pub async fn export_sbom(
    format: String,
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let content =
        use_case::export_sbom(format, &inventory.assets).map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &content).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn export_stix_bundle(output_path: String, state: &AppState) -> Result<String, AppError> {
    let capture = read_state(&state.capture, "capture").map_err(AppError::state_lock)?;
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let json = use_case::export_stix_bundle(&inventory.assets, &capture.connections)
        .map_err(AppError::invalid_input)?;
    write_text_file(std::path::Path::new(&output_path), &json).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn save_topology_image(
    image_data: String,
    output_path: String,
) -> Result<String, AppError> {
    use_case::save_topology_image(image_data, output_path).map_err(AppError::invalid_input)
}

pub async fn export_filtered_pcap(
    filter_ips: Vec<String>,
    filter_ports: Vec<u16>,
    output_path: String,
    state: &AppState,
) -> Result<FilteredPcapResult, AppError> {
    let input_paths = read_state(&state.capture, "capture")
        .map_err(AppError::state_lock)?
        .imported_files
        .clone();
    use_case::export_filtered_pcap(&input_paths, filter_ips, filter_ports, output_path)
        .map_err(AppError::invalid_input)
}

pub async fn generate_communication_allowlist(
    state: &AppState,
) -> Result<Vec<AllowlistEntry>, AppError> {
    let capture = read_state(&state.capture, "capture").map_err(AppError::state_lock)?;
    let inventory = read_state(&state.inventory, "inventory").map_err(AppError::state_lock)?;
    let analysis = read_state(&state.analysis, "analysis").map_err(AppError::state_lock)?;
    Ok(use_case::generate_communication_allowlist(
        &inventory.assets,
        &capture.connections,
        &analysis.connection_stats,
    ))
}

pub async fn export_allowlist_csv(
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let entries = generate_communication_allowlist(state).await?;
    let csv = use_case::export_allowlist_csv(&entries);
    write_text_file(std::path::Path::new(&output_path), &csv).map_err(AppError::IoError)?;
    Ok(output_path)
}

pub async fn export_firewall_rules(
    output_path: String,
    state: &AppState,
) -> Result<String, AppError> {
    let entries = generate_communication_allowlist(state).await?;
    let rules = use_case::export_firewall_rules(&entries);
    write_text_file(std::path::Path::new(&output_path), &rules).map_err(AppError::IoError)?;
    Ok(output_path)
}
